tabular/src/autogluon/tabular/utils.py

```diff
@@ -150,3 +156,222 @@ def infer_eval_metric(problem_type: str) -> Scorer:
+            final_iteration = i + num_features_processing == feature_count
+            if (num_features_processing < compute_count) and final_iteration:
+                X_raw = X_raw.loc[:row_count * num_features_processing - 1]
+
+            row_index = 0
+            for feature in parallel_computed_features:
+                row_index_end = row_index + row_count
+                X_raw.loc[row_index:row_index_end - 1, feature] = X_shuffled[feature].values
+                row_index = row_index_end
+
+            X_raw_transformed = X_raw if transform_func is None else transform_func(X_raw, **transform_func_kwargs)
+            y_pred = predict_func(X_raw_transformed, **predict_func_kwargs)
+
+            row_index = 0
+            for feature in parallel_computed_features:
+                # calculating importance score for given feature
+                row_index_end = row_index + row_count
+                y_pred_cur = y_pred[row_index:row_index_end]
+                score = eval_metric(y, y_pred_cur)
+                fi[feature] = score_baseline - score
+
+                if not final_iteration:
+                    # resetting to original values for processed feature
+                    X_raw.loc[row_index:row_index_end - 1, feature] = X[feature].values
+
+                row_index = row_index_end
+        fi_dict_list.append(fi)
+        shuffle_repeats_completed = shuffle_repeat + 1
+        if time_limit is not None and shuffle_repeat != (num_shuffle_sets - 1):
+            time_now = time.time()
+            time_left = time_limit - (time_now - time_start)
+            time_permutation_average = (time_now - time_permutation_start) / (shuffle_repeat + 1)
+            if time_left < (time_permutation_average * 1.1):
+                if not silent:
+                    logger.log(20, f'\tEarly stopping feature importance calculation before all shuffle sets have completed due to lack of time...')
+                break
+    fi = dict()
+    fi_stddev_dict = dict()
+    fi_z_score_dict = dict()
+    for feature in features:
+        fi_fold_feature_list = [permutation_importance_dict_repeat[feature] for permutation_importance_dict_repeat in fi_dict_list]
+        fi[feature] = np.mean(fi_fold_feature_list)
+        if len(fi_fold_feature_list) > 1:
+            fi_stddev_dict[feature] = np.std(fi_fold_feature_list, ddof=1)
+        else:
+            fi_stddev_dict[feature] = np.nan
+        if fi_stddev_dict[feature] != np.nan and fi_stddev_dict[feature] != 0:
+            fi_z_score_dict[feature] = fi[feature] / fi_stddev_dict[feature]
+        elif fi_stddev_dict[feature] != np.nan:  # stddev = 0
+            if fi[feature] == 0:
+                fi_z_score_dict[feature] = np.nan
+            elif fi[feature] > 0:
+                fi_z_score_dict[feature] = np.inf
+            else:  # < 0
+                fi_z_score_dict[feature] = -np.inf
+        else:
+            fi_z_score_dict[feature] = np.nan
```

Comment:
possible duplicate - this looks very similar to bagged_ensemble_model.py: 308-322

Reply:
modularized and avoided code duplication